### PR TITLE
[Search] Add ID prefix matching

### PR DESCRIPTION
**Pull request by loic-sharma**
# ID Prefix Matching

This change adds package ID prefix matching to our search service. For example, searching for `ent` should return Entity Framework results. The implementation has the following behaviors:

*  Prefix match only the last term. The query `foo bar` matches result `barn` but not result `food`
* If the last term "looks" like a package ID (ie: contains package ID separators), prefix match on entire package ID values. The query `foo.bar` matches package ID `foo.barXXX` but not `XXXfoo.bar`
* ... Otherwise, prefix match on package ID tokens. The query `bar` matches result `foo.barXXX`
* Boost prefix matching if the last term has fewer than 4 characters
* Disable prefix matching if there are any field scoped terms (like `owner:billy`)

These behaviors were chosen through experimentation. You can find my prototypes here: https://github.com/NuGet/NuGetGallery/issues/7128#issuecomment-561886058

Part of https://github.com/NuGet/NuGetGallery/issues/7128
Part of https://github.com/NuGet/NuGetGallery/issues/7764

## Offline results

Data set | Control NDCG score | Treatment NDCG score | Result
-- | -- | -- | --
nuget.org curated queries | 0.8526 | 0.8523 | ❌
client curated queries | 0.7627 | 0.8215 | ✔
feedback queries | 0.7316 | 0.7708 | ✔

<details>
    <summary>Search scorer detailed results</summary>

Curated Search Queries
======================
Control:   0.8526305050062
Treatment: 0.851433027547545

Biggest Winners (10)
--------------------
entityframework           => +0.0007
system.windows.forms      => +0.0007
aws                       => +0.0007
System.ServiceModel       => +0.0006
identityserver            => +0.0004
xml                       => +0.0004
Microsoft.AspNet.WebPages => +0.0003
powershell                => +0.0003
Microsoft.IdentityModel   => +0.0002
itext7                    => +0.0002

Biggest Losers (10)
-------------------
excel                          => -0.0017
kafka                          => -0.0007
Microsoft.AspNet.WebApi.Client => -0.0006
swagger                        => -0.0005
soap                           => -0.0004
postgres                       => -0.0004
email                          => -0.0003
roslyn                         => -0.0003
AspNetCore.HealthChecks        => -0.0003
Excel                          => -0.0002

Lowest Treatment Scores (10)
----------------------------
system.web.mvc      => 0
System.Web.Mvc      => 0
system.web.http     => 0
email               => 0
System.Web          => 0
Tags:"seq-app"      => 0
Tags:"excel-to-pdf" => 0
socket              => 0
httpcontext         => 0
qrcode              => 0

Client Curated Search Queries
=============================
Control:   0.762687841979449
Treatment: 0.821543990654414

Biggest Winners (10)
--------------------
m      => +0.0057
mic    => +0.0052
mi     => +0.0044
Mic    => +0.0039
enti   => +0.0037
ent    => +0.0033
M      => +0.0029
swa    => +0.0028
entit  => +0.0028
micros => +0.0026

Biggest Losers (10)
-------------------
swagger                        => -0.0009
ef                             => -0.0009
excel                          => -0.0008
Microsoft.AspNet.WebApi.Client => -0.0006
net                            => -0.0004
nu                             => -0.0003
MySql.Data                     => -0.0002
identity                       => -0.0001
Identity                       => 0
web                            => 0

Lowest Treatment Scores (10)
----------------------------
micro           => 0
Micro           => 0
System.Web      => 0
boot            => 0
c               => 0
system.web.http => 0
se              => 0
ef              => 0
test            => +0.0003
logging         => +0.0004

Feedback
========
Control:   0.731618985150309
Treatment: 0.76471068648356

Biggest Winners (10)
--------------------
appsf                           => +0.0117
mpart                           => +0.0085
microsoft.visualstudio.services => +0.0057
MS, .net core                   => +0.0039
Azure                           => +0.0024
brot                            => +0.0024
clrheap                         => +0.0024
eventgri                        => +0.0024
fluentassertion                 => +0.0024
logging                         => +0.0024

Biggest Losers (10)
-------------------
Microsoft.Extensions => -0.0142
ef                   => -0.0024
Ef                   => -0.0024
extended xml         => -0.0024
rx                   => -0.0024
microsoft.azure      => -0.0020
ORM                  => -0.0018
postgres             => -0.0018
Postgres             => -0.0018
Android support      => -0.0014

Lowest Treatment Scores (10)
----------------------------
log file                                    => 0
codeanalysis.csharp                         => 0
log                                         => 0
ef                                          => 0
Vault                                       => 0
extended xml                                => 0
alexa                                       => 0
Microsoft.Extensions.Hosting.WindowsService => 0
mariadb                                     => 0
nsq                                         => 0
</details>

<details>
<summary>Prefix boosting alternatives</summary>

nuget.org curated control: 0.8526305050062
nuget.org curated treatment @ 1: 0.851433027547545
nuget.org curated treatment @ 10: 0.852301307516017
nuget.org curated treatment @ 20: 0.852301307516016

client curated control: 0.762687841979449
client curated treatment @ 1: 0.821543990654414
client curated treatment @ 10: 0.82191663050274
client curated treatment @ 20: 0.821468599623034

feedback control: 0.731618985150309
feedback treatment @ 1: 0.76471068648356
feedback treatment @ 10: 0.770140204780968
feedback treatment @ 20: 0.770798417984278
</details>

### CODE REVIEW

**Comment by joelverhagen** [comment]
For the `Lowest Treatment Scores (10)` that look like prefixes but still have score 0, what are the results? Do they look okay? Perhaps the data set needs another pair of eyes in such cases.

**Comment by joelverhagen** [comment]
Could we have a single, very black and white test put in the relevancy functional tests for this behavior?

File: src/NuGet.Services.AzureSearch/SearchService/SearchTextBuilder.cs [hunk 1]
```diff
@@ -190,6 +190,34 @@ public string Build(ParsedQuery parsed)
...
+                        builder.AppendScopedTerm(
+                            fieldName: IndexFields.PackageId,
+                            term: lastUnscopedTerm,
+                            required: false,
+                            prefixSearch: true,
+                            boost: _options.Value.PrefixMatchBoost);
+                    }
+                    else
+                    {
+                        var boost = lastUnscopedTerm.Length < 4
```
Comment (joelverhagen) [inline src/NuGet.Services.AzureSearch/SearchService/SearchTextBuilder.cs]: How did we pick 4?

Comment (loic-sharma) [reply src/NuGet.Services.AzureSearch/SearchService/SearchTextBuilder.cs]: This was chosen through experimentation. It looks like I didn't record these experiments' results, I'll redo them and update the issue.

**Comment by loic-sharma** [comment]
🚧 This comment is still a WIP

> For the Lowest Treatment Scores (10) that look like prefixes but still have score 0, what are the results? Do they look okay? Perhaps the data set needs another pair of eyes in such cases.

Query | Expected (curated data set) | Actual
------- | ----------------------------- | -------
nu | NUnit, NUnit3TestAdapter, NuGet.*, Nunit.* | NuGet.Frameworks, NuGet.CommandLine, System.Runtime.Numerics, NuGet.Versioning, NuGet.Common
micro | Microsoft.Extensions.*, Microsoft.AspNetCore.*, Microsoft.AspNet.* ,Microsoft.Owin.*, Microsoft.* | Dapper, Caliburn.Micro, MicroCHAP, Caliburn.Micro.Core, MicroLite
boot | bootstrap, Twitter.Bootstrap, Angular.UI.Bootstrap, Bootstrap.datepicker, Bootstrap.less | Nancy.BootStrappers.StructureMap, Boot.MultiTenancy, Tempest.Boot, DisUtils.BootConfig
c | Castle.Core, CommonServiceLocator, CsvHelper, Common.Logging, Microsoft.Extensions.Configuration | Microsoft.AspNetCore.Hosting.Abstractions, Microsoft.AspNetCore.Hosting.Server.Abstractions, Microsoft.CSharp, Microsoft.AspNetCore.Mvc.Core
se | Serilog, Selenium.*, SendGrid, ServiceStack.*, Serilog.* | Microsoft.AspNetCore.Hosting.Server.Abstractions, System.Runtime.CompilerServices.Unsafe, System.Security.Principal.Windows, System.Security.Cryptography.X509Certificates, System.Security.Cryptography.Algorithms

⚠️ alex in feedback data set should have a result of Alexa.Net, looks like a copy paste error

File: src/NuGet.Services.AzureSearch/SearchService/SearchTextBuilder.cs (hunk 1 above)
Comment (loic-sharma) [reply src/NuGet.Services.AzureSearch/SearchService/SearchTextBuilder.cs]: Test | nuget.org curated score | client curated score | feedback score
-- | -- | -- | --
No boost | ... | ... | ..
Always boost | ... | ... | ..
Boost if query length < 1 | ... | ... | ..
Boost if query length < 2 | ... | ... | ..
Boost if query length < 3 | ... | ... | ..
Boost if query length < 4 | 0.849712852684465 | 0.820892078212731 | 0.778338146528021
Boost if query length < 5 | ... | ... | ..
Boost if query length < 6 | ... | ... | ..
Boost if query length < 7 | ... | ... | ..

**Pull request merged**